### PR TITLE
Add guidelines for code contributions

### PR DESCRIPTION
**Pull request by adnapibar**
Signed-off-by: Rabi Panda <adnapibar@gmail.com>

### Description
A set of guidelines for contributors to help them deciding whether a feature should be included in OpenSearch.
 
### Issues Resolved
N/A 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dreamer-89 — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -59,6 +59,16 @@ If you would like to contribute to the documentation, please do so in the [docum
 
 As with other types of contributions, the first step is to [**open an issue on GitHub**](https://github.com/opensearch-project/OpenSearch/issues/new/choose). Opening an issue before you make changes makes sure that someone else isn't already working on that particular problem. It also lets us all work together to find the right approach before you spend a bunch of time on a PR. So again, when in doubt, open an issue.
 
+Additionally, here are a few guidelines to help you decide whether a particular feature should be included in OpenSearch,
+
+**Is your feature important to most users of OpenSearch?**
+
+If you believe that your feature is going to fulfill the need of most users of OpenSearch, then this should belong in OpenSearch. However, we don't want every feature to be built into the server, so if the feature requires additional permissions or bring in extra dependencies it should be included as a core module.
```
Comment (mch2) [inline CONTRIBUTING.md:66]:
```suggestion
If you believe that your feature is going to fulfill a need for most users of OpenSearch, then this belongs in OpenSearch. However, we don't want every feature to be built into the server, so if the feature requires additional permissions or brings in extra dependencies it should instead be included as a core module.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -59,6 +59,16 @@ If you would like to contribute to the documentation, please do so in the [docum
...
+
+**Is your feature a common dependency across multiple plugins?**
+
+Does this feature contain functionality that cuts across multiple plugins? If so, then most likely, this should belong in OpenSearch as a core module or plugin.
```
Comment (mch2) [inline CONTRIBUTING.md:70]:
```suggestion
Does this feature contain functionality that cuts across multiple plugins? If so, this most likely belongs in OpenSearch as a core module or plugin.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -59,6 +59,16 @@ If you would like to contribute to the documentation, please do so in the [docum
 
 As with other types of contributions, the first step is to [**open an issue on GitHub**](https://github.com/opensearch-project/OpenSearch/issues/new/choose). Opening an issue before you make changes makes sure that someone else isn't already working on that particular problem. It also lets us all work together to find the right approach before you spend a bunch of time on a PR. So again, when in doubt, open an issue.
 
+Additionally, here are a few guidelines to help you decide whether a particular feature should be included in OpenSearch,
```
Comment (mch2) [inline CONTRIBUTING.md:62]:
```suggestion
Additionally, here are a few guidelines to help you decide whether a particular feature should be included in OpenSearch.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -59,6 +59,16 @@ If you would like to contribute to the documentation, please do so in the [docum
 
 As with other types of contributions, the first step is to [**open an issue on GitHub**](https://github.com/opensearch-project/OpenSearch/issues/new/choose). Opening an issue before you make changes makes sure that someone else isn't already working on that particular problem. It also lets us all work together to find the right approach before you spend a bunch of time on a PR. So again, when in doubt, open an issue.
 
+Additionally, here are a few guidelines to help you decide whether a particular feature should be included in OpenSearch.
+
+**Is your feature important to most users of OpenSearch?**
+
+If you believe that your feature is going to fulfill a need for most users of OpenSearch, then this belongs in OpenSearch. However, we don't want every feature to be built into the server, so if the feature requires additional permissions or brings in extra dependencies it should instead be included as a core module.
+
+**Is your feature a common dependency across multiple plugins?**
```
Comment (elfisher) [inline CONTRIBUTING.md:68]: I like this a lot. I think this also is a good criteria for being an extension point too.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -59,6 +59,16 @@ If you would like to contribute to the documentation, please do so in the [docum
 
 As with other types of contributions, the first step is to [**open an issue on GitHub**](https://github.com/opensearch-project/OpenSearch/issues/new/choose). Opening an issue before you make changes makes sure that someone else isn't already working on that particular problem. It also lets us all work together to find the right approach before you spend a bunch of time on a PR. So again, when in doubt, open an issue.
 
+Additionally, here are a few guidelines to help you decide whether a particular feature should be included in OpenSearch.
+
+**Is your feature important to most users of OpenSearch?**
+
+If you believe that your feature is going to fulfill a need for most users of OpenSearch, then this belongs in OpenSearch. However, we don't want every feature to be built into the server, so if the feature requires additional permissions or brings in extra dependencies it should instead be included as a core module.
```
Comment (elfisher) [inline CONTRIBUTING.md:66]: Minor wording suggestions:

If you believe that a feature is going to fulfill a need for most users of OpenSearch, then it belongs in OpenSearch. However, we don't want every feature built into the core server. If the feature requires additional permissions or brings in extra dependencies it should instead be included as a module in core.

**Review by elfisher — APPROVED** [review]
This lgtm. I've left some minor wording suggestions.

**Review by dblock — COMMENTED** [review]
I like it. I would also say something about whether the feature can have alternate implementations. For example, for job-scheduler we ask whether job-scheduler be included in core -> will we ever want two different implementations of job-scheduler. If the answer were a definite yes, we'd not want it in core.

**Comment by nknize** [comment]
> ...will we ever want two different implementations of job-scheduler. If the answer were a definite yes, we'd not want it in core.

I don't use this criteria. Snapshot repository modules, for example, have multiple implementations in core (S3, Azure, GCS). They're in the core as modules because we want them installed as part of the min default experience. I could see two different job schedulers running in parallel using different implementations (cron vs task).

**Review by nknize — APPROVED** [review]

**Comment by navneet1v** [comment]
Does this change encourages that as the feature owner we should create plugins or move to a module within the OpenSearch repo? This is bit unclear here. what is the thought process for the same?

**Comment by CEHENKLE** [comment]
We want things to get built in the place makes the most sense for them to be built :)  

From my POV, it's less about encouraging features owners to build in one place or the other, and more about making sure they're thinking about what is the best place to build things using a lens of commonality.  

Does that makes sense @navneet1v ?

**Comment by navneet1v** [comment]
> We want things to get built in the place makes the most sense for them to be built :)
> 
> From my POV, it's less about encouraging features owners to build in one place or the other, and more about making sure they're thinking about what is the best place to build things using a lens of commonality.
> 
> Does that makes sense @navneet1v ?

Yes it make sense. Thanks for providing the clarity.

**Pull request merged**